### PR TITLE
fix(auth): hasRole handles when currentUser.roles is a string

### PR DESCRIPTION
**Pull request by dac09**
We had an edge case where we had missing logic in hasRoles.

The updated code now handles the case where:
- currentUser.roles is a string
- but hasRoles is called with an array of strings

### CODE REVIEW

**Comment by dac09** [comment]
@Tobbe mind taking a look please? This is something Rob and I already fixed on the dbAuth template

**Comment by dac09** [comment]
Related #4489

**Comment by dthyresson** [comment]
@dac09 Does this fix https://github.com/redwoodjs/redwood/issues/4488 or https://github.com/redwoodjs/redwood/issues/4489

I am almost done with https://github.com/redwoodjs/redwood/issues/4492 so the directive can take a string or an array of string roles.

One changge I was making to auth was from:

```
export const hasRole = ({ roles }: { roles: AllowedRoles }): boolean => {
```

to

```
export const hasRole = ( roles : AllowedRoles ): boolean => {
```

so one could use `hasRole('bazinga')` instead of `hasRole(roles: 'bazinga')` per @cannikin

**Comment by dac09** [comment]
It doesn't fix those issues DT (which is why I haven't linked them) - this fixes a different edge case we don't handle, as explained in the description

File: packages/auth/src/__tests__/AuthProvider.test.tsx
Comment (Tobbe) [inline packages/auth/src/__tests__/AuthProvider.test.tsx:21]:
```suggestion
  roles?: Array<string> | string
```
Changed to match `CurrentUser` in AuthProvider

**Comment by dthyresson** [comment]
> 

Ok just checking. Thanks,

File: packages/auth/src/AuthProvider.tsx
Comment (Tobbe) [inline packages/auth/src/AuthProvider.tsx:196]: Don't we need an `isArray` check on `currentUser.roles` here as well? As you added below

File: packages/auth/src/__tests__/AuthProvider.test.tsx
Comment (dac09) [reply packages/auth/src/__tests__/AuthProvider.test.tsx:21]: equivalent no?

File: packages/auth/src/AuthProvider.tsx
Comment (dac09) [reply packages/auth/src/AuthProvider.tsx:196]: Ummm in theory, if you passed a function or something, but we assume that you will pass either a string or an array, and don't do exhaustive checks for that. Think thats ok!

File: packages/auth/src/__tests__/AuthProvider.test.tsx
Comment (Tobbe) [reply packages/auth/src/__tests__/AuthProvider.test.tsx:21]: 100%, so why not do the same in both places?
Personally, for whatever reason, I always do `string[]`, but I prefer `Record<string, string>`. Don't know why I like the utility helper for one of them but not the other...

File: packages/auth/src/AuthProvider.tsx
Comment (Tobbe) [reply packages/auth/src/AuthProvider.tsx:196]: What I'm saying is if `roleToCheck` is a string, and `this.state.currentUser.roles` is a string we should not do `includes(roleToCheck)` but rather use `===`.

Comment (Tobbe) [reply packages/auth/src/AuthProvider.tsx:196]: Unless there's something I don't understand about the code here....

Comment (cannikin) [reply packages/auth/src/AuthProvider.tsx:196]: That's true, if you had `admin_user` and `user`, it'd return `true` in both cases if you checked for the role `user` with `includes()` 😬

Comment (cannikin) [reply packages/auth/src/AuthProvider.tsx:196]: So with the current setup there's 4 cases it needs to handle:

1. currentUser.roles is a string, roleToCheck is a string
2. currentUser.roles is an array, roleToCheck is a string
3. currentUser.roles is a string, roleToCheck is an array
4. currentUser.roles is an array, roleToCheck is an array

I feel like we're only covering 3 of these in here aren't we?

Comment (cannikin) [reply packages/auth/src/AuthProvider.tsx:196]: That check on lines 195-196 would cause the partial string problem if `currentUser.roles` is itself a string, but if it changes to `===` then it'll fail if currentUser.roles. is an array.

Comment (Tobbe) [reply packages/auth/src/AuthProvider.tsx:196]: Exactly, we have 4 cases it needs to handle, but only three branches in the code. I'd add another one to (explicitly) handle the missing case too

Comment (dac09) [reply packages/auth/src/AuthProvider.tsx]: @cannikin, @Tobbe how about this one?

File: packages/cli/src/commands/setup/auth/templates/dbAuth.auth.ts.template
Comment (dac09) [inline packages/cli/src/commands/setup/auth/templates/dbAuth.auth.ts.template:83]: Note: has role now returns false when none of the conditions are met. Previously we were returning true - presumably because we were always calling hasRoles from requireAuth, even when roles weren't supplied by the user. 

in Line 105, I now add a check to only run `hasRole` if roles are supplied

**Comment by dac09** [comment]
Ok gents, I've updated all the templates!

File: tasks/test-project/tasks.js
Comment (dac09) [inline tasks/test-project/tasks.js:298]: This is to prevent typescript complaining on the smoke tests. TS knows from Primsa types that currentUser.roles can never be a string[] in the test project, so part of the code gets marked as an error.

File: __fixtures__/test-project/api/src/functions/auth.ts
Comment (Tobbe) [inline __fixtures__/test-project/api/src/functions/auth.ts:142]: Why is this cookie thing included in this PR?

File: __fixtures__/test-project/api/src/lib/auth.ts
Comment (Tobbe) [inline __fixtures__/test-project/api/src/lib/auth.ts:60]: The variable is named `roles`, but the comment says `roleToCheck`. Should we align those?

File: packages/auth/src/AuthProvider.tsx
Comment (Tobbe) [inline packages/auth/src/AuthProvider.tsx:190]: According to the types `rolesToCheck` can never be `undefined`. Is this code for the benefit of JS projects? Maybe add a comment about that if so. Easy for someone to see this code in the future and not think about that and then remove this check...

Also, could we make this an early return to limit the nesting here a bit.

File: packages/cli/src/commands/setup/auth/templates/auth.ts.template
Comment (Tobbe) [inline packages/cli/src/commands/setup/auth/templates/auth.ts.template:78]: As above, `roles` and `roleToCheck` don't match

File: __fixtures__/test-project/api/src/lib/auth.ts
Comment (Tobbe) [inline __fixtures__/test-project/api/src/lib/auth.ts:56]: Why is the typecasting here needed?

**Review by Tobbe — CHANGES_REQUESTED** [review]

File: __fixtures__/test-project/api/src/lib/auth.ts
Comment (dac09) [reply __fixtures__/test-project/api/src/lib/auth.ts:60]: Intentionally done, don't want to change directives, etc. and the destructure syntax to rename looks bad

File: packages/auth/src/AuthProvider.tsx
Comment (dac09) [reply packages/auth/src/AuthProvider.tsx:190]: This is for the benefit of JS projects, yes. I am trying to change as little as possible, so I keep any extra checks that are there.

On early return:
Note that I've changed the code below in requireAuth, where hasRoles is called. hasRoles is never called when roles isn't defined, so the check is a bit redundant I suppose

File: __fixtures__/test-project/api/src/lib/auth.ts
Comment (dac09) [reply __fixtures__/test-project/api/src/lib/auth.ts:56]: See here: https://github.com/redwoodjs/redwood/pull/4678/files#r822434312

File: __fixtures__/test-project/api/src/functions/auth.ts
Comment (dac09) [reply __fixtures__/test-project/api/src/functions/auth.ts:142]: Because the fixture wasn't regenerated before

File: __fixtures__/test-project/api/src/lib/auth.ts
Comment (dac09) [reply __fixtures__/test-project/api/src/lib/auth.ts:60]: Also I don't think I should change the comment to match the variable name. It makes it cleaarer what the purpose of roles is.

Comment (Tobbe) [reply __fixtures__/test-project/api/src/lib/auth.ts:60]: I don't agree. To me this looks confusing. Like "What's `roleToCheck`? I don't see anything with that name."
How about a comment explaining what `roles` is, and then keep using `roles` everywhere?

Like

```
// `roles` is the role, or roles, that the user needs to have at least one of
// for this check to be successful. So we check for these roles against
// `currentUser.roles`
```

Maybe that turned out too wordy, but you get the idea...

**Comment by dac09** [comment]
If you look at the auth file, it's littered with long comments everywhere. I will change it to roles, if you really want, because I don't think its worth either our time debating this, and it adds 0-0.1% more value.

File: packages/auth/src/AuthProvider.tsx
Comment (Tobbe) [reply packages/auth/src/AuthProvider.tsx:190]: > so the check is a bit redundant I suppose

Hehe. So remove it? Or keep it and explain why in a comment?

File: __fixtures__/test-project/api/src/lib/auth.ts
Comment (Tobbe) [reply __fixtures__/test-project/api/src/lib/auth.ts:56]: Ohh, sorry, I didn't know how all this worked.
So tasks/test-project/tasks.js is used to generate this fixture, right? That's why this line changed?

Comment (dac09) [reply __fixtures__/test-project/api/src/lib/auth.ts:56]: Yup that's right. You use the `--rebuild-fixture` flag on test project, and it'll update the fixture with the latest template changes, codemods, etc!

File: packages/auth/src/AuthProvider.tsx
Comment (dac09) [reply packages/auth/src/AuthProvider.tsx:190]: I'm hesitant to remove this, because requireAuth and hasRoles are used in so many different contexts that I can't be sure! What would you suggest to explain in the comment that isn't obvious in the code already?

Comment (Tobbe) [reply packages/auth/src/AuthProvider.tsx:190]: Taking a second look at the code, it should be fine to just remove the check, right?
Let's say `rolesToCheck` is `undefined`.
`if (typeof rolesToCheck === 'string')` will be false. And so will `if (Array.isArray(rolesToCheck))`. So it will fall through to `return false` anyway.

That said, if you do want to keep it I'd do something like this
```
    // Because this code is used from JS projects as well, we can't rely on the
    // typechecking of `rolesToCheck` to protect us from `undefined`
    if (typeof rolesToCheck === 'undefined' || !this.state.currentUser?.roles) {
      return false
    }
```

Comment (dac09) [reply packages/auth/src/AuthProvider.tsx:190]: Ok removed, I noticed this is just in auth provider, not the src/lib/auth one. Fine by me!

File: __fixtures__/test-project/api/src/lib/auth.ts
Comment (dac09) [reply __fixtures__/test-project/api/src/lib/auth.ts:60]: Ok I found a nice compromise. Watcha think? 

I really dislike long wordy comments in this file, especially not next to the code... we just have so much info on there!

Comment (Tobbe) [reply __fixtures__/test-project/api/src/lib/auth.ts:60]: > I really dislike long wordy comments in this file, especially not next to the code... we just have so much info on there!

Agreed! But maybe it really is mostly helpful for people the first time they see it? And they might not generate new RW projects as often? I mean, I'm just not sure if I'm more annoyed by it because I generate new RW apps almost daily to test some thing 😅  But this is probably a discussion for some other time.

I'll have a look at the compromise now!

Comment (Tobbe) [reply __fixtures__/test-project/api/src/lib/auth.ts:60]: I like it!

**Review by Tobbe — APPROVED** [review]

**Comment by dac09** [comment]
I'll wait for @cannikin to do a once over, and then merge!

**Review by cannikin — APPROVED** [review]

**Comment by thedavidprice** [comment]
Merging this to continue work on #4673

Thanks, all!

**Pull request merged**
